.github/workflows/build-and-test.yml

```diff
@@ -0,0 +1,102 @@
+---
+
+name: Build and Test
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+    branches: [ "main" ]
+env:
+  IMAGE_PREFIX: "ghcr.io/${{ github.repository }}"
+  REGISTRY: 'ghcr.io'
+
+jobs:
+  build-test-push:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        include:
+          - workdir: microservice_a
+          - workdir: microservice_b
+          - workdir: microservice_c
+      fail-fast: false
+    permissions:
+      contents: read
+      packages: write
+    steps:
+      - name: checkout
+        uses: actions/checkout@v3
+
+      - name: "build ${{matrix.workdir}} image"
+        working-directory: "${{matrix.workdir}}"
+        run: "./build.sh ${{github.sha}}"
+
+      - name: "run ${{matrix.workdir}} container"
+        working-directory: "${{matrix.workdir}}"
+        run: >-
+          docker run
+          --name ${{matrix.workdir}}__${{github.sha}}
+          -d
+          -P
+          ${{ env.IMAGE_PREFIX }}/${{matrix.workdir}}:${{github.sha}}
+
+      - name: "cache | pip"
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip
+          restore-keys: |
+            ${{ runner.os }}-pip-
+
+      - name: "test | docker healthcheck"
+        working-directory: "${{matrix.workdir}}"
+        run: |
```

Comment:
wait for it for a reasonable time and get this random tcp port instead of hardcoding it